clash-prelude/tests/Clash/Tests/BitVector.hs

```diff
@@ -1,19 +1,51 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
 {-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
 
-module Clash.Tests.BitVector where
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.Extra.Solver #-}
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.Normalise #-}
+{-# OPTIONS_GHC -fplugin=GHC.TypeLits.KnownNat.Solver #-}
+
+module Clash.Tests.BitVector (tests, main) where
+
+import qualified Hedgehog as H
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
 
 import Data.Proxy
 import GHC.TypeNats (KnownNat, SomeNat (..), natVal, someNatVal)
 import Test.Tasty
 import Test.Tasty.HUnit
-import Test.Tasty.QuickCheck
 
-import Clash.Prelude (bitPattern)
+import qualified Test.Tasty.Hedgehog as H
+import qualified Test.Tasty.QuickCheck as Q
+
+import Clash.Prelude
+  (Bit, high, low, bitPattern, type (<=), type (-), natToInteger, liftA2, msb)
 import Clash.Sized.Internal.BitVector (BitVector (..))
 
 import Clash.Tests.SizedNum
 
+-- | Generates a BitVector either with the MSB set or not, and checks whether
+-- 'msb' agrees with it.
+msbTest :: forall n. (1 <= n, KnownNat n) => H.Property
+msbTest = H.property $ do
+  (bv, b) <- H.forAll msbGen
+  b H.=== msb bv
```

Comment:
This catches the problem by the virtue of the definition of `Eq Bit`, which indeed compares full values in the datatype. For instance `Show Bit` would not catch the problem because a malformed `Bit` might print correctly.

I think the way of testing used here is fine. I do think it's important to realise that it works because `Eq Bit` correctly compares malformed `Bit`s, which it is obviously not required to do as no malformed `Bit`s should ever occur in the first place.

Reply:
On the other hand, even if `Eq Bit` did a less thorough comparison, I suppose the chances that all randomly generated values happen to coincide requires a really odd bug.